File: units.js
class OptimizedUnits {
    constructor(units) {
        this.units = units;
        this.dp = Array.from({ length: 100 }, () => Array(101).fill(-1));
    }

    getCoins(amt, i, units) {
        if (amt === 0) return 0;
        if (amt < 0 || i >= units.length) return 1e9;

        if (this.dp[i][amt] !== -1) return this.dp[i][amt];

        let take = 1e9;
        const sol = this.getCoins(amt - units[i], i, units);
        if (sol !== 1e9) take = 1 + sol;

        let skip = this.getCoins(amt, i + 1, units);

        return this.dp[i][amt] = Math.min(take, skip);
    }

    findAverage() {
        let numCoins = 0;
        for (let amt = 1; amt <= 99; amt++) {
            numCoins += this.getCoins(amt, 0, this.units);
        }
        this.dp = Array.from({ length: 100 }, () => Array(101).fill(-1));
        return numCoins / 99;
    }
}

const units = [1, 2, 5, 10, 20, 50];
const res = new OptimizedUnits(units);
console.log("Average coins used:", res.findAverage().toFixed(4)); // Output: 3.4343

File: units.cpp
#include <bits/stdc++.h>

using namespace std;

class OptimizedUnits
{
private:
    vector<int> units;
    vector<vector<int>> dp;

public:
    OptimizedUnits(vector<int> &units)
    {
        this->units = units;
        dp = vector<vector<int>>(100, vector<int>(101, -1));
    }

    int getCoins(int amt, int i, vector<int> &units)
    {
        if (amt == 0)
            return 0;
        if (amt < 0 || i >= units.size())
            return INT_MAX;

        if (dp[i][amt] != -1)
            return dp[i][amt];

        int take = INT_MAX;
        int sol = getCoins(amt - units[i], i, units);
        if (sol != INT_MAX && sol + 1 >= 0)
            take = sol + 1;

        int skip = getCoins(amt, i + 1, units);

        return dp[i][amt] = min(take, skip);
    }

    double findAverage()
    {
        int numCoins = 0;
        for (int amt = 1; amt <= 99; ++amt)
        {
            int result = getCoins(amt, 0, units);
            numCoins += (result == INT_MAX) ? 0 : result;
        }
        dp = vector<vector<int>>(100, vector<int>(101, -1));
        return static_cast<double>(numCoins) / 99;
    }
};

int main()
{
    vector<int> units = {1, 2, 5, 10, 20, 50};
    OptimizedUnits res(units);
    cout << fixed << setprecision(4);
    cout << "Average coins used: " << res.findAverage() << endl;
    return 0;
}
